### PR TITLE
Build Electron from source

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yu-re-ka** [comment]
Updated to 23.1.1, fixed the issue with "Regenerating ninja files" loop so it actually builds now

**Comment by yu-re-ka** [comment]
unsure if we want to package multiple versions of electron from source, or just fall back to binaries when an application _requires_ an older electron version?

**Comment by dotlambda** [comment]
> unsure if we want to package multiple versions of electron from source, or just fall back to binaries when an application _requires_ an older electron version?

I think we should package multiple versions. Packages often only work with one of them and it would be nice to have as many packages as possible profit from electron built from source. Would the expressions for the different versions of electron differ a lot?

**Comment by prusnak** [comment]
I don't thing we should be messing with electron packages right now.

Let's make the `electron-oss` package first (containing electron build from source).
And only after testing we can rename `electron` to `electron-bin` and  alias `electron-oss` to `electron`.

**Comment by SuperSandro2000** [comment]
> I don't thing we should be messing with `electron` packages at this stage.

What makes you think that? Shouldn't the output be the same? Why shouldn't it be?



> And only after testing we can rename `electron` to `electron-bin` and alias `electron-oss` to `electron` in another PR.

If the from source build works as expected we can just delete the bin variant. Or for which usecase do we need both?

File: pkgs/development/tools/electron/default.nix [hunk 1]
```diff
@@ -1,171 +1,96 @@
...
 
 let
-  mkElectron = import ./generic.nix args;
+  info = lib.importJSON ./info.json;
 in
-rec {
 
-  electron = electron_23;
+(chromium.override { channel = "electron_23"; }).mkDerivation (base: rec {
+  name = "electron";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/default.nix:8]:
```suggestion
  pname = "electron";
```

File: pkgs/development/tools/electron/default.nix [hunk 2]
```diff
@@ -1,171 +1,96 @@
...
+      src = ./version.patch;
+      inherit version;
+    })
+  ];
+
+  postUnpack = ''
+    mv chromium-* src
+    sourceRoot=src
+  '' + (lib.concatStrings (lib.mapAttrsToList (path: dep: ''
+    mkdir -p ${builtins.dirOf path}; cp -r ${fetchFromGitHub dep} ${path}; chmod u+w -R ${path}
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/default.nix:32]:
```suggestion
    mkdir -p ${builtins.dirOf path}
    cp -r ${fetchFromGitHub dep} ${path}
    chmod u+w -R ${path}
```

File: pkgs/development/tools/electron/default.nix [hunk 3]
```diff
@@ -1,171 +1,96 @@
...
+  '') info.deps)) + ''
+    (
+      cd src/electron
+      export HOME=$NIX_BUILD_TOP/fake_home
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/default.nix:36]:
```suggestion
      export HOME=$TMPDIR/fake_home
```
NIX_BUILD_TOP should be avoided

File: pkgs/development/tools/electron/default.nix [hunk 4]
```diff
@@ -1,171 +1,96 @@
...
-
-  electron_20 = mkElectron "20.3.11" {
-    armv7l-linux = "709b9eb958e9488f6375811041179556b9cd0b8fc1eab6b899ef4a89423f98b2";
-    aarch64-linux = "0f488ac9eeda2baa4c4e571fd75ac8e055dac9dcdf83051164232b1005a29224";
-    x86_64-linux = "7899bf391ae35e10d78a5da622e506dd4ae859cd8c18953cd2dc54f1a5e5225e";
-    x86_64-darwin = "751204887aa553c2a7811d3cb04d71e85359ccce2cf21d38e43eda24575ef4db";
-    aarch64-darwin = "8ea1a446b41413b97d83d2955a4800c5f7c9061662f78c3e8d96827741f8e211";
-    headers = "06s4z2hs9sbri4jsjrgybq0sn7rrx7zf3iwfg8da1wb6ahwqcd7w";
-  };
+  postFixup = " ";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/default.nix:94]: Not sure what the proper way to clear this is. Maybe set to null?

File: pkgs/development/tools/electron/default.nix [hunk 5]
```diff
@@ -1,171 +1,96 @@
...
+  '') info.deps)) + ''
+    (
+      cd src/electron
+      export HOME=$NIX_BUILD_TOP/fake_home
+      yarn config --offline set yarn-offline-mirror $yarnOfflineCache
+      fixup_yarn_lock yarn.lock
+      yarn install --offline --frozen-lockfile --ignore-scripts --no-progress --non-interactive
+    )
+
+    (
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/default.nix:42]: I am a bit unsure how well error handling works in subshells. Does this properly abort if a patch cannot be applied?

**Comment by prusnak** [comment]
> What makes you think that? Shouldn't the output be the same? Why shouldn't it be?

It's easier to review smaller non-intrusive changes first. The doing the rest later.

Most of the stuff in this PR is refactoring the existing electron packages which makes the whole review harder and takes away the focus from reviewing the important stuff.

File: pkgs/development/tools/electron/default.nix (hunk 1 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/default.nix:8]: this is specific to chromium.mkDerivation

**Comment by prusnak** [comment]
Since https://github.com/NixOS/nixpkgs/pull/218608 has been merged, can you please rebase this PR on top of master?

**Comment by yu-re-ka** [comment]
Will do later. I'm still trying to figure out what to do about older electron releases. Packaging fix releases for the previous release channels is much more complicated, because the underlying chromium  release channels don't receive fix/security updates anymore and thus there are no release tarballs for the chromium revisions.

Electron 20.3.12 uses Chromium 104.0.5112.124 while the last release is 104.0.5112.101
Electron 21.4.2 uses Chromium 106.0.5249.199 while the last release is 106.0.5249.119
Electron 22.3.1 uses Chromium 108.0.5359.215 while the last release is 108.0.5359.124

- Only package the latest Electron version from source => we already said that is not really useful because lots of applications depend on a specific older version of Electron
- Package older Electron releases with mismatching chromium revisions => potentially bad if people think they have the latest Electron fix release, but are actually missing important chromium security fixes that were never properly released in this chromium major version
- Use gclient in a fixed-output derivation to fetch the chromium source code for unreleased revisions
- Use an approach like [robotnix](https://github.com/danielfullmer/robotnix/blob/master/apks/chromium/mk-vendor-file.py) to generate a nix expression from the gclient DEPS file and obtain the exact versions of all dependencies individually in many smaller derivations

**Comment by yu-re-ka** [comment]
For now I have opted for option 2 and packaged each electron release with the latest available chromium release that is matching three out of four parts version tuple.

I will add a warning to the nix expression when the packaged chromium version is older than the one used by upstream.

This can be improved in a followup PR.

File: pkgs/development/tools/electron/default.nix (hunk 5 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/default.nix]: The build does abort if a patch can not be applied

```
patching src/third_party/squirrel.mac/vendor/Mantle
patching src/third_party/squirrel.mac/vendor/ReactiveObjC
patching src/third_party/angle
patching src/third_party/boringssl/src
patching src
error: src/content/test/data/accessibility/mac/selection/selectall-textarea-expected.txt: No such file or directory
/nix/store/yc6qxp5g68aly1k237dr3hkkm2rsdfba-stdenv-linux/setup: line 136: pop_var_context: head of shell_variables not a function context

File: pkgs/development/tools/electron/default.nix (hunk 3 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/default.nix]: done

File: pkgs/development/tools/electron/default.nix (hunk 1 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/default.nix]: changed to "packageName"

File: pkgs/development/tools/electron/default.nix (hunk 4 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/default.nix]: obsolete, the postFixup section from the chromium derivation now applies to electron as well

**Comment by yu-re-ka** [comment]
> Since #218608 has been merged, can you please rebase this PR on top of master?

Done

With the latest commit on this branch, I have now built electron-source.electron_{21,22,23} successfully

**Comment by SuperSandro2000** [comment]
> 3\. Use gclient in a fixed-output derivation to fetch the chromium source code for unreleased revisions

That probably takes to long to build.

I would say it is fine for now to build all from source where it is easily possible and fall back to binary if the chromium version is way to old and causes to much trouble.

**Comment by yu-re-ka** [comment]
> That probably takes to long to build.

I have tinkered with gclient in the past to make it do shallow clones, which reduced the initial sync time to less than ten minutes with a gigabit downlink.

This would have to be re-packed into a tar.xz though because the output exceeds Hydra size limits otherwise. We had this with qtwebengine in the past.

**Comment by yu-re-ka** [comment]
I did some work on the "generate a nix expression from the gclient DEPS file" option.

Sadly, cipd (chromium infrastructure package deployment) works differently than I thought. Generally in cipd, a package plus tag/version is resolved to a package "instance" which is according to the documentation "a ZIP file with the package file contents", and the instance id is the zipfile hash. I thought the update script could resolve the package version to an instance (zipfile hash) and its download url and then generate a Nix fetchzip, _but_ the urls that the cipd registry hands out have an expiry date, so the cipd registry has to be queried for a fresh link each time a package is downloaded, _so_ we need a fetchcipd fetcher.

**Comment by yu-re-ka** [comment]
The source-built electron binaries will also need gtk/gsettings wrapping, though I'm not sure if that should be done using wrapGAppsHook or manually (as chromium does), and how it fits in with the binary electron wrapping (can they share one wrapper?).

**Comment by yu-re-ka** [comment]
implemented "generate a nix expression from the gclient DEPS file" solution.
Just leaving out the cipd dependencies seems to work, since they are mostly build tool binaries (ninja, gn, nodejs, jdk) which build ourselves.

File: pkgs/development/tools/electron/info.json [hunk 6]
```diff
@@ -0,0 +1,3290 @@
...
+                "gn": {
+                    "version": "2022-12-12",
+                    "url": "https://gn.googlesource.com/gn",
+                    "rev": "5e19d2fb166fbd4f6f32147fbb2f497091a54ad8",
+                    "sha256": "1b5fwldfmkkbpp5x63n1dxv0nc965hphc8rm8ah7zg44zscm9z30"
+                }
+            }
+        }
+    }
+}
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/info.json:3290]: editorconfig check complains that a new line is missing here

File: pkgs/development/tools/electron/update.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+from datetime import datetime
+from urllib.request import urlopen
+
+depot_tools_checkout = tempfile.TemporaryDirectory()
+subprocess.check_call([
+    "nix-prefetch-git",
+    "--builder", "--quiet",
+    "--url", "https://chromium.googlesource.com/chromium/tools/depot_tools",
+    "--out", depot_tools_checkout.name,
+    "--rev", "543fdf32d15fe47a0beadc3f5a3e0d6218ad237c"])
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/update.py:23]: Is that a hardcoded rev?

File: pkgs/development/tools/electron/common.nix [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    echo '#define GPU_LISTS_VERSION "${info.deps."src".rev}"'             >> gpu/config/gpu_lists_version.h
+    echo '#endif  // GPU_CONFIG_GPU_LISTS_VERSION_H_'                     >> gpu/config/gpu_lists_version.h
+
+    echo '/* Generated by lastchange.py, do not edit.*/'                   > skia/ext/skia_commit_hash.h
+    echo '#ifndef SKIA_EXT_SKIA_COMMIT_HASH_H_'                           >> skia/ext/skia_commit_hash.h
+    echo '#define SKIA_EXT_SKIA_COMMIT_HASH_H_'                           >> skia/ext/skia_commit_hash.h
+    echo '#define SKIA_COMMIT_HASH "${info.deps."src/third_party/skia".rev}-"' >> skia/ext/skia_commit_hash.h
+    echo '#endif  // SKIA_EXT_SKIA_COMMIT_HASH_H_'                        >> skia/ext/skia_commit_hash.h
+
+    echo -n '${info.deps."src/third_party/dawn".rev}'                         > gpu/webgpu/DAWN_VERSION
```
Comment (SuperSandro2000) [inline pkgs/development/tools/electron/common.nix:97]:
```suggestion
    echo -n '${info.deps."src/third_party/dawn".rev}'                     > gpu/webgpu/DAWN_VERSION
```

File: pkgs/development/tools/electron/update.py (hunk 7 above)
Comment (yu-re-ka) [reply pkgs/development/tools/electron/update.py:23]: Yes, it is a known working rev. But should be independent of the chromium version.

**Comment by yu-re-ka** [comment]
built electron-source.electron_22 and electron-source.electron_25

**Pull request closed without merging**